### PR TITLE
Fix `_should_reload_dl_epoch` causing inconsistent validation dataloader reloading

### LINKED ISSUE #10948: Inconsistent validation dataloader reloading when using `reload_dataloaders_every_n_epochs` and `check_val_every_n_epoch` flags
## 🐛 Bug
When the `reload_dataloaders_every_n_epochs` and `check_val_every_n_epoch` flags of the Trainer are used, the validation dataloader may reload inconsistently or not reload at all.

A few examples:

When `reload_dataloaders_every_n_epochs` = 2 and `check_val_every_n_epoch` = 3, the validation dataloader reloads every 6 epochs (should be 3).

When `reload_dataloaders_every_n_epochs` = 2 and `check_val_every_n_epoch` = 2, the validation dataloader **never reloads**.

### To Reproduce

[BoringModel Demo](https://colab.research.google.com/drive/165HJ5lYW4RMub1RLGDhy2TRHmcrPQfyC?usp=sharing)

### Expected behavior

When `reload_dataloaders_every_n_epochs`<= `check_val_every_n_epoch`, I expect the validation dataloader to reload before each validation run. 

### Environment
* CUDA:
	- GPU:
	- available:         False
	- version:           11.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.10.0+cu111
	- pytorch-lightning: 1.5.4
	- tqdm:              4.62.3
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.12
	- version:           # 1 SMP Sat Jun 5 09:50:34 PDT 2021

### Additional context

The culprit is probably [`_should_reload_dl_epoch`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#LC1843). The logic `not self.current_epoch % n_epochs` doesn't consider when the validation dataloader was last reloaded.

cc @borda @justusschock @awaelchli @ninginthecloud @tchaton

### PR DESCRIPTION
**Pull request by adamviola**
## What does this PR do?

Fixes #10948

When the flag `reload_dataloaders_every_n_epochs` is set, PyTorch Lightning reloads dataloaders on epochs that are multiples of `reload_dataloaders_every_n_epochs`. This causes inconsistent validation dataloader reloading when the flags`check_val_every_n_epoch` or `val_check_interval` are set.

This PR now has the trainer store the epoch of the last reload of the train and validation dataloaders. Each dataloader is now eligible to reload when `current_epoch` - `dl_last_reload_epoch` >= `reload_dataloaders_every_n_epochs`.

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Actually yes

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -1119,17 +1119,12 @@ def test_dataloaders_load_only_once(tmpdir):
...
 
     # logger file to get meta
     trainer = Trainer(
-        default_root_dir=tmpdir,
-        limit_train_batches=10,
-        limit_val_batches=10,
-        val_check_interval=0.3,
-        reload_dataloaders_every_n_epochs=1,
-        max_epochs=3,
+        default_root_dir=tmpdir, limit_train_batches=0.3, limit_val_batches=0.3, num_sanity_val_steps=0, max_epochs=3
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:1127]: Why change the values for `limit_*_batches?`

Also I'd prefer keeping the the previous `num_sanity_val_steps` value

**Review by carmocca — APPROVED** [review]
Great work!

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (adamviola) [reply tests/trainer/test_dataloaders.py:1127]: This diff is misleading. Check out the [changes for the second commit](https://github.com/PyTorchLightning/pytorch-lightning/pull/11036/commits/123984d9a1ec72aaf34dd6cd58892c054b96b6ac).

I'm removing the `test_dataloaders_load_only_once_val_interval` test because I realized that the test I added in the first commit [`test_dataloaders_load_every_n_epochs_frequent_val`](https://github.com/PyTorchLightning/pytorch-lightning/pull/11036/commits/4c818369208c2bc5be0fca6f7dd86fa7e82229d0#diff-c439da4427c1b40699495f578c0619586bde3992b3a57aeef80fefe2fa940c1aR1266) checks the same case.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1354,6 +1358,14 @@ def _run_sanity_check(self, ref_model):
             # restore the previous stage when the sanity check if finished
             self.state.stage = stage
 
+    def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -> None:
+        self._last_train_dl_reload_epoch = self.current_epoch
+        return super().reset_train_dataloader(model=model)
+
+    def reset_val_dataloader(self, model: Optional["pl.LightningModule"] = None) -> None:
+        self._last_val_dl_reload_epoch = self.current_epoch
+        return super().reset_val_dataloader(model=model)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1367]: In the context of #10417, which proposes to move the data loading related methods and attributes to the data class, I think these changes should go to `data_loading.py`(where the reset method are).

Comment (adamviola) [reply pytorch_lightning/trainer/trainer.py]: Sounds good! I moved the changes to `data_loading.py`

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -1075,44 +1069,108 @@ def test_dataloaders_load_only_once_no_sanity_check(tmpdir):
...
+    )
+    model.test_dataloader = Mock(wraps=model.test_dataloader)
+
+    trainer.fit(model)
+    trainer.test(model)
+
+    # Verify epoch of reloads
+    if n == 3:
+        assert train_reload_epochs == [0, 2, 4, 6, 8]
+        assert val_reload_epochs == [0, 2, 5, 8]
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1124]: shouldnt this go into the parametrization (the expected sequence)?
then there is no check needed here.

File: tests/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -1075,44 +1069,108 @@ def test_dataloaders_load_only_once_no_sanity_check(tmpdir):
...
+    if n == 5:
+        assert train_reload_epochs == [0, 2, 4, 6, 8]
+        assert val_reload_epochs == [0, 4, 9]
+
+    # Sanity check at epoch 0 creates a validation dataloader
+    # but val is checked every n epochs starting from epoch n-1
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1130]: this comment is floating by itself, where does it belong?

File: tests/trainer/test_dataloaders.py [hunk 5]
```diff
@@ -1075,44 +1069,108 @@ def test_dataloaders_load_only_once_no_sanity_check(tmpdir):
...
+    elif n == 2:
+        assert train_reload_epochs == [0, 2, 4]
+        assert val_reload_epochs == [0, 2, 4]
 
-@pytest.mark.parametrize("n", [1, 2])
-def test_dataloaders_load_every_n_epochs(tmpdir, n):
-    model = BoringModel()
+
+@pytest.mark.parametrize("n", [3, 5])
+def test_dataloaders_load_every_n_epochs_infrequent_val(tmpdir, n):
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1094]: can you add a small docstring to these tests explaining in one sentence what is being tested?

File: tests/trainer/test_dataloaders.py [hunk 6]
```diff
@@ -1159,15 +1217,6 @@ def validation_step(self, batch, batch_idx):
...
         call.val_dataloader(),
-        # This has subsequent calls to val_dataloader
-        # because the training loop runs the evaluation loop,
-        # which reloads the val dataloader again.
-        # We cannot yet rely on trainer.current_epoch=0 to skip reloading
-        # the val dataloader on the first epoch because this only tracks the training epoch
-        # meaning multiple passes through the validation data within a single training epoch
-        # would not have the dataloader reloaded.
-        # This breaks the assumption behind reload_dataloaders_every_n_epochs=1
-        call.val_dataloader(),
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1170]: really nice that this can be fixed now!!

File: tests/trainer/test_dataloaders.py (hunk 3 above)
Comment (adamviola) [reply tests/trainer/test_dataloaders.py:1124]: Not exactly sure what you mean here. Are you suggesting to add another arg to the parametrize decorator with the expected output? Or compute what the expected sequence should be using `n`?

File: pytorch_lightning/trainer/data_loading.py [hunk 7]
```diff
@@ -471,3 +481,15 @@ def _check_eval_shuffling(dataloader, mode):
                 " it is strongly recommended that you turn this off for val/test/predict dataloaders.",
                 category=PossibleUserWarning,
             )
+
+    @property
+    def _should_reload_train_dl(self) -> bool:
+        """Check if train dataloader should be reloaded."""
+        n_epochs = self.reload_dataloaders_every_n_epochs
+        return n_epochs and self.current_epoch - self._last_train_dl_reload_epoch >= n_epochs
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:489]:
```suggestion
        return n_epochs and (self.current_epoch - self._last_train_dl_reload_epoch >= n_epochs)
```

File: pytorch_lightning/trainer/data_loading.py [hunk 8]
```diff
@@ -471,3 +481,15 @@ def _check_eval_shuffling(dataloader, mode):
...
+
+    @property
+    def _should_reload_val_dl(self) -> bool:
+        """Check if validation dataloader should be reloaded."""
+        n_epochs = self.reload_dataloaders_every_n_epochs
+        return n_epochs and self.current_epoch - self._last_val_dl_reload_epoch >= n_epochs
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:495]:
```suggestion
        return n_epochs and (self.current_epoch - self._last_val_dl_reload_epoch >= n_epochs)
```

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1020]: Can we merge this test with the one below ?

File: tests/trainer/test_dataloaders.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1124]: You could do the following

```py
@pytest.mark.parametrize("n, train_reload_epochs_expect, val_reload_epochs", [
   (3, [0, 2, 4, 6, 8], [0, 2, 5, 8]),
   (5, [0, 2, 4, 6, 8], [0, 4, 9])])
```

**Review by tchaton — APPROVED** [review]
Really nice contribution!

File: pytorch_lightning/trainer/data_loading.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:489]: This is purely to improve readability.

**Review by Borda — APPROVED** [review]

File: tests/trainer/test_dataloaders.py (hunk 5 above)
Comment (adamviola) [reply tests/trainer/test_dataloaders.py]: Added!

File: tests/trainer/test_dataloaders.py (hunk 4 above)
Comment (adamviola) [reply tests/trainer/test_dataloaders.py]: Just explaining the expected validation dataloader reload behavior. Moved it to a better spot.

**Review by awaelchli — APPROVED** [review]
